tests/test_mypy.py

```diff
@@ -0,0 +1,164 @@
+"""This test file is ignored by mypy since we're using pytest-mypy-testing to
+assert the mypy errors that should be raised.
+
+Assertions in mypy are done via these line comments:
+    # N: <msg> - assert a mypy note message
+    # W: <msg> - assert a mypy warning message
+    # E: <msg> - assert a mypy error message
+    # R: <msg> - assert a mypy note message Revealed type is '<msg>'.
+
+NOTE:
+    Pytest's parametrization doesn't work here due to the dynamic nature of
+    variables in each line. For-loops also doesn't work well since the type
+    checker needs to narrow down on the type within the test function's scope.
+"""
+
+import pytest
+from web_poet import HttpResponse
+
+from zyte_common_items import (
+    Breadcrumb,
+    Image,
+    Link,
+    Product,
+    ProductFromList,
+    ProductList,
+    ProductVariant,
+)
+
+response = HttpResponse("https://www.some-page/", b"")
+
+
+@pytest.mark.mypy_testing
+def test_assignment_image():
+    img = Image(url=response.url)
+    img.url = response.url
+    img.url = str(response.url)
+    img.url = 123  # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+
+@pytest.mark.mypy_testing
+def test_assignment_breadcrumb():
+    bc = Breadcrumb(url=response.url)
+    bc.url = response.url
+    bc.url = str(response.url)
+    bc.url = 123  # E: Incompatible types in assignment (expression has type "int", variable has type "Optional[str]")
+
+
+@pytest.mark.mypy_testing
+def test_assignment_link():
+    link = Link(url=response.url)
+    link.url = response.url
+    link.url = str(response.url)
+    link.url = 123  # E: Incompatible types in assignment (expression has type "int", variable has type "Optional[str]")
```

Comment:
This might only fail because of a previous assignment, not because of annotation of Link objects. I.e. the following code fails mypy checks

```
x = "hello"
x = 5
# Incompatible types in assignment (expression has type "int", variable has type "str")
```

Reply:
Ahhh nice catch! I used distinct variables instead to avoid this. https://github.com/zytedata/zyte-common-items/pull/11/commits/57859876b8f9eca829d202f8fd48e97a3e19f298